### PR TITLE
BUG: `Styler.apply` consistently manages Series return objects aligning labels.

### LINKED ISSUE #13657: Row based style.apply doesn't modify background colors
#### Code Sample, a copy-pastable example if possible

``` py
# add color gradient to column(s)
import numpy as np
import math
from pandas import *
from IPython.display import display, HTML

# color gradient
COLORS = ["#F1F2B5", "#CCD7A5", "#A7BC96", "#82A186", "#5D8677", "#386B67", "#135058"]

# rows/columns to skip applying gradient to
SKIPROWSORCOLUMNS = ["Day","Users"] 

# index in row or columns to skip
SKIPINDICES = [0,1]

# dataframe to operate on
df = DataFrame([["Monday",100,.65,.39,.33],["Tuesday",120,.45,.32,.31],["Wednesday",150,.75,.33,.30]])
df.columns = ["Day","Users","M1","M2","M3"] #column headers

# takes dataframe and colors
# returns colormap
def applyColorGradient(data, colors=COLORS, skiprowsorcolumns=SKIPROWSORCOLUMNS, skipindices=SKIPINDICES):
  numberOfColors = len(colors)
  i = 0
  maxNumberUnset = True
  maxNumber = 0
  while (i < len(data)):
    if (i not in skipindices):
      if(maxNumberUnset):
        maxNumber = data[i]
        maxNumberUnset = False
      else:
        current = data[i]
        if(current > maxNumber):
          maxNumber = current
    i = i + 1
  name = data.name
  dontskip = data.name not in skiprowsorcolumns
  colorArray = [colors[int(math.floor(value/maxNumber*(numberOfColors-1)))] if dontskip and index not in skipindices else '' for index,value in enumerate(data)]
  print(colorArray)
  colorSeries = pandas.Series(["background-color:" + color if color else '' for color in colorArray], name=data.name)
  display(colorSeries)
  return colorSeries


# applying colorgradient to dataframe
gradientdf = df.style.apply(applyColorGradient, axis=1) #0 is used for columns

# display table
display(gradientdf) #0 is used for columns



```
#### Expected Output

With axis = 0, this manages to give each cell a background color that is indicated. When I change this to a row based function, this doesn't color any of the cells, even though the color series are coming in. Is this a bug or am I doing something wrong?
## INSTALLED VERSIONS

commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-91-generic
machine: x86_64
processor: 
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8

pandas: 0.17.1
nose: None
pip: 8.1.2
setuptools: 23.0.0
Cython: 0.23.5
numpy: 1.10.4
scipy: 0.17.1
statsmodels: 0.6.1
IPython: 4.2.0
sphinx: None
patsy: 0.4.1
dateutil: 2.5.2
pytz: 2016.3
blosc: None
bottleneck: None
tables: None
numexpr: 2.5.2
matplotlib: 1.5.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.13
pymysql: None
psycopg2: None
Jinja2: None

### PR DESCRIPTION
**Pull request by attack68**
- [x] closes #13657 

This issue is due to the asymmetric application of `DataFrame.apply` (#42005). 
**The bug** is that when using `Styler.apply` the treatment of Series return objects is applied for `axis=0` and overwritten for `axis=1`.

```python
df = pd.DataFrame([[1,2],[3,4]], index=["X", "Y"], columns=["X", "Y"])
s = df.style

def series_ret(s):
    return pd.Series(["color:red;", "color:blue;"], index=["Y", "X"])  # note X-Y order
```

| Command    | Current Behaviour             | New Behaviour            |
| ------------ | ------------------------- | ---------------------- |
| `s.apply(series_ret, axis=0)` | ![][image5] | ![][image5] |
| `s.apply(series_ret, axis=1)` | ![][image6] | ![][image7] |

Also the application has been expanded to **allow** (for axis=0 | 1) Series return objects that are **not** the same size, or (for axis=None) DataFrame return objects that are **not** the same size, if they contain only relevant index labels. This allows for more flexible user defined functions. Potentially it also increases performance by not having to loop over labels that no longer exist. 

```python
def series_ret(s):
    return pd.Series(["color:red;"], index=["Y"])  # note only returning 1 element

def array_ret(s):
    return ["color:red;"]

def ser_no_idx_ret(s):
    return pd.Series(["color:red;", "color:blue;"]) # note returning a Series of right shape with no index
```

| Command    | Current Behaviour             | New Behaviour            |
| ------------ | ------------------------- | ---------------------- |
| `s.apply(series_ret, axis=0)` | `ValueError: Function returned the wrong shape.` | ![][image1] |
| `s.apply(series_ret, axis=1)` | `ValueError: Length mismatch: Expected axis has 1 elements, new values have 2 elements` | ![][image2] |
| `s.apply(array_ret, axis=0)` | `ValueError: Function returned the wrong shape.` | `ValueError: Function created invalid index labels. Usually, this is the result of the function returning a Series which contains invalid labels, or returning an incorrectly shaped, list-like object which cannot be mapped to labels, possibly due to applying the function along the wrong axis.` |
| `s.apply(array_ret, axis=1)` |`ValueError: Length mismatch: Expected axis has 1 elements, new values have 2 elements`                                |`ValueError: Function created invalid index labels. Usually, this is the result of the function returning a Series which contains invalid labels, or returning an incorrectly shaped, list-like object which cannot be mapped to labels, possibly due to applying the function along the wrong axis.` |
| `s.apply(ser_no_idx_ret, axis=0)` | `KeyError: 0` | `ValueError: Function created invalid index labels. Usually, this is the result of the function returning a Series which contains invalid labels, or returning an incorrectly shaped, list-like object which cannot be mapped to labels, possibly due to applying the function along the wrong axis.` |
| `s.apply(ser_no_idx_ret, axis=1)` | ![][image6] |`ValueError: Function created invalid index labels. Usually, this is the result of the function returning a Series which contains invalid labels, or returning an incorrectly shaped, list-like object which cannot be mapped to labels, possibly due to applying the function along the wrong axis.` |

Note all errors do report the shapes, I just didn't include in the table for brevity.

### Edge Cases

The edge case here is that if a user has a DataFrame with a default range index (or a DataFrame with some indexes the same as those in default, e.g. 0 or 1) then even returning arrays of the wrong size may work, since `DataFrame.apply` assigns them to a RangeIndex.

To be consistent with the above this should raise, but it is impossible to hook into the return value of the user-defined function and determine if it is a Series (where non-matching shapes are OK) or list-like (where matching shapes should be enforced)

```python
df = pd.DataFrame([[1,2],[3,4]])
s = df.style
```
| Command    | Current Behaviour             | New Behaviour            |
| ------------ | ------------------------- | ---------------------- |
| `s.apply(array_ret, axis=0)` | `ValueError: Function returned the wrong shape.` | ![][image3] |
| `s.apply(array_ret, axis=1)` | `ValueError: Length mismatch: Expected axis has 1 elements, new values have 2 elements` |  ![][image4] |

### ASV
Benchmarks unchanged.

[image1]: https://user-images.githubusercontent.com/24256554/122002332-e59b5580-cdb1-11eb-85c2-993c7e20bdfc.png
[image2]: https://user-images.githubusercontent.com/24256554/122002344-e9c77300-cdb1-11eb-8a01-f50ec660b300.png
[image3]: https://user-images.githubusercontent.com/24256554/122004954-57c16980-cdb5-11eb-8ba4-ec14b1a6b06f.png
[image4]: https://user-images.githubusercontent.com/24256554/122005013-69a30c80-cdb5-11eb-900c-0d4be813cb58.png
[image5]: https://user-images.githubusercontent.com/24256554/122006522-3497b980-cdb7-11eb-892e-f34c80e3c7ff.png
[image6]: https://user-images.githubusercontent.com/24256554/122006532-36fa1380-cdb7-11eb-8b08-f89d6991536c.png
[image7]: https://user-images.githubusercontent.com/24256554/122006816-93f5c980-cdb7-11eb-8aff-09b658c594d9.png

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm small comments

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -1125,9 +1125,10 @@ def _apply(
         subset = slice(None) if subset is None else subset
         subset = non_reducing_slice(subset)
         data = self.data.loc[subset]
-        if axis is not None:
-            result = data.apply(func, axis=axis, result_type="expand", **kwargs)
-            result.columns = data.columns
+        if axis in [0, "index"]:
```
[the remark this answers is not in the document]
Comment (attack68) [reply pandas/io/formats/style.py:1128]: issue added #43048

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -1175,14 +1185,17 @@ def apply(
...
+            valid index labels considering ``subset``.
             ``func`` should take a DataFrame if ``axis`` is ``None`` and return either
-            an ndarray with the same shape or a DataFrame with identical columns and
-            index.
+            an ndarray with the same shape or a DataFrame, not necessarily of the same
+            shape, with valid index and columns labels considering ``subset``.
 
             .. versionchanged:: 1.3.0
 
+            .. versionchanged:: 1.4.0
```
[the remark this answers is not in the document]
Comment (attack68) [reply pandas/io/formats/style.py:1197]: added

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @attack68

**Pull request merged**